### PR TITLE
Add unit tests for nncf quantization functions

### PR DESCRIPTION
**Pull request by Candyzorua**
### Details

- Added unit tests for quantization and accuracy control functions
- Files tested: 
    * [nncf/quantization/algorithms/accuracy_control/evaluator.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/evaluator.py)
    * [nncf/quantization/algorithms/accuracy_control/subset_selection.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/subset_selection.py)
    * [nncf/quantization/algorithms/accuracy_control/openvino_backend.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/openvino_backend.py)
    * [nncf/quantization/algorithms/accuracy_control/ranker.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/ranker.py)
    
    * [nncf/quantization/algorithms/accuracy_control/algorithm.py](https://app.codecov.io/gh/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/algorithm.py)   
- For [ranker.py](https://github.com/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/ranker.py), I am not sure how to test the case with multithreading, so I have left that out for now.
- For [algorithm.py](https://github.com/openvinotoolkit/nncf/blob/develop/nncf/quantization/algorithms/accuracy_control/algorithm.py), I am not sure how to test some of the more complex functions like `apply`, so I have left that out for now.
- Addresses https://github.com/openvinotoolkit/nncf/issues/2496

### CODE REVIEW

**Comment by andrey-churkin** [comment]
@Candyzorua Thanks for the contribution! Some general comments from my side.

It is not allowed to import backend-specific parts of code in common tests. For example, it is not permitted to import `LinearModel` from the `tests.openvino.native.models` in the `test_evaluator.py` module. Typically, backend-specific modules have the name of the backend (backend is a name of a framework) in the path.

To understand how to resolve this problem, please see how the `TemplateTestNNCFTensorOperators` class is used. It would be great to implement something like this here.

Please feel free to ask another questions for further explanation.

**Comment by Candyzorua** [comment]
Thanks @andrey-churkin ! Will get it resolved asap.

**Comment by andrey-churkin** [comment]
@Candyzorua Please let me know if you need any help. Thanks.

**Comment by Candyzorua** [comment]
@andrey-churkin Thanks! Just a few clarifications:

By  "implement something like `TemplateTestNNCFTensorOperators`, do you mean I should focus on testing the mathematical correctness of the algorithms?

Also, I believe that not importing backend-specific parts of code would make it impossible to test certain functions, such as the function to retrieve the backend-specific PreparedModel. As such, I think it may only be possible to achieve around 50-60% coverage on some files. Would that be acceptable?

**Comment by Candyzorua** [comment]
@andrey-churkin all references to backend-specific code should be removed :)

**Comment by Candyzorua** [comment]
Thank you for your comments @andrey-churkin! I have addressed them.

For `get_subset_indices_pot_version()`, I noticed that the original implementation does not produce the same output as  `get_subset_indices_pot_version()` because they have different methods of breaking ties. I have made a change to `get_subset_indices_pot_version()`, so it produces the same output.

Please let me know if the previous implementation is preferred and I will rollback the change + edit the test case. 

Thank you!

**Comment by andrey-churkin** [comment]
@Candyzorua Thanks! Yes, the difference between `get_subset_indices_pot_version()` and `get_subset_indices()` is expected. So I suggest to take the same as in `test_get_subset_indices()` `errors` and `subset_size` values for `test_get_subset_indices_pot_version()` test, but change only `expected_indices`.

File: tests/common/accuracy_control/test_algorithm.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        ),
+        NNCFNode(
+            {
+                NNCFNode.NODE_NAME_ATTR: "node_name_2",
+            }
+        ),
+    ]
+    result = _create_message(nodes)
+    assert isinstance(result, str)
+    assert result == "\n".join(["\t" + node.node_name for node in nodes])
```
Comment (andrey-churkin) [inline tests/common/accuracy_control/test_algorithm.py:42]: Try to obtain expected values without adding any additional code wherever possible. It allows for quickly understanding the expected behavior of the method being tested.
```suggestion
    assert result == "\tnode_name_1\n\tnode_name_2"
```

File: tests/common/accuracy_control/test_algorithm.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    report.removed_groups = [type("MockGroup", (), {"quantizers": [node1], "operations": [node2]})]
+    return report
+
+
+def test_removed_quantizers(setup_quantization_accuracy_restorer_report):
+    assert setup_quantization_accuracy_restorer_report.removed_quantizers[0].node_name == "node_name_1"
+
+
+def test_reverted_operations(setup_quantization_accuracy_restorer_report):
+    assert setup_quantization_accuracy_restorer_report.reverted_operations[0].node_name == "node_name_2"
```
Comment (andrey-churkin) [inline tests/common/accuracy_control/test_algorithm.py:67]: I suggest representing the test in the following way:

```python
@pytest.mark.parametrize(
    # test data here
)
def test_quantization_accuracy_restorer_report_properties(
    removed_groups,
    expected_removed_quantizers,
    expected_reverted_operations
):
    report = QuantizationAccuracyRestorerReport()
    report.removed_groups = removed_groups
    assert report.removed_quantizers == expected_removed_quantizers
    assert report.expected_reverted_operations == expected_reverted_operations
    
```
It allows us to quickly understand what the input data for the test is and what the expected values are.

**Comment by Candyzorua** [comment]
Thanks again for the guidance @andrey-churkin. I have addressed your comments and implemented your proposed changes.

**Review by andrey-churkin — APPROVED** [review]
@Candyzorua Please run pre-commit checks to be sure that all modified files correspond code style. Thanks!

```
pip install  black==23.3.0 isort==5.12.0 pre-commit==3.2.2
wget https://raw.githubusercontent.com/openvinotoolkit/nncf/develop/.isort.cfg -O .isort.cfg
wget https://raw.githubusercontent.com/openvinotoolkit/nncf/develop/.pre-commit-config.yaml -O .pre-commit-config.yaml
pre-commit run --files <CHANGED_FILES>

```

**Pull request merged**